### PR TITLE
WIP: Clamav

### PR DESCRIPTION
**Pull request by pweil-**
Support clamAV as a scan type

This is super rough but working on a preliminary level:

```
[pweil@localhost image-inspector]$ sudo _output/local/bin/local/image-inspector --scan-type clamav --image fedora:22 --path /tmp/clamd.socket
&{stream: Eicar-Test-Signature FOUND Eicar-Test-Signature stream  0 FOUND} <nil>
```

### CODE REVIEW

**Comment by pweil-** [comment]
@mfojtik @legionus

File: pkg/inspector/image-inspector.go [hunk 1]
```diff
@@ -99,46 +100,58 @@ func NewDefaultImageInspector(opts iicmd.ImageInspectorOptions) ImageInspector {
...
+	var scanReport []byte
+	var htmlScanReport []byte
+	var scanner iiapi.Scanner
 
-	if err = i.pullImage(client); err != nil {
-		return err
-	}
+	switch i.opts.ScanType {
+	case "openscap":
+		client, err := docker.NewClient(i.opts.URI)
```
Comment (legionus) [inline pkg/inspector/image-inspector.go]: @pweil- Maybe it would be better to move this to a separate openscap-specific source ?

**Comment by simon3z** [comment]
Nice! cc @enoodle @ilackarms

File: pkg/inspector/image-inspector.go (hunk 1 above)
Comment (enoodle) [inline pkg/inspector/image-inspector.go]: @pweil- I think that you have moved the downloading of the image / serving the results under only `openscap` scanning by accident.

File: pkg/api/types.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ type APIVersions struct {
 // Scanner interface that all scanners should define.
 type Scanner interface {
 	// Scan will scan the image
-	Scan(string, *docker.Image) error
+	Scan(path string, image *docker.Image) error
```
Comment (mfojtik) [inline pkg/api/types.go]: maybe just a nit but I think we want something like *ScanOptions here so we can extended it in the future with more options?

File: pkg/inspector/image-inspector.go [hunk 3]
```diff
@@ -99,46 +100,58 @@ func NewDefaultImageInspector(opts iicmd.ImageInspectorOptions) ImageInspector {
 
 // Inspect inspects and serves the image based on the ImageInspectorOptions.
 func (i *defaultImageInspector) Inspect() error {
-	client, err := docker.NewClient(i.opts.URI)
-	if err != nil {
-		return fmt.Errorf("Unable to connect to docker daemon: %v\n", err)
-	}
+	var scanReport []byte
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: group this?

File: pkg/inspector/image-inspector.go [hunk 4]
```diff
@@ -99,46 +100,58 @@ func NewDefaultImageInspector(opts iicmd.ImageInspectorOptions) ImageInspector {
...
+	var htmlScanReport []byte
+	var scanner iiapi.Scanner
 
-	if err = i.pullImage(client); err != nil {
-		return err
-	}
+	switch i.opts.ScanType {
+	case "openscap":
```
Comment (mfojtik) [inline pkg/inspector/image-inspector.go]: should be a const?

File: pkg/api/types.go (hunk 2 above)
Comment (enoodle) [inline pkg/api/types.go]: We should also probably have an output struct for the scan results.

File: pkg/api/types.go [hunk 5]
```diff
@@ -27,7 +27,7 @@ func (osm *OpenSCAPMetadata) SetError(err error) {
 }
 
 var (
-	ScanOptions = []string{"openscap"}
+	ScanOptions = []string{"openscap", "clamav"}
```
Comment (mfojtik) [inline pkg/api/types.go]: Scanners ?

Comment (mfojtik) [inline pkg/api/types.go]: or ScanType

File: pkg/clamav/scanner.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+func NewScanner(socket string) api.Scanner {
+	return &ClamScanner{
+		Socket: socket,
+	}
+}
+
+// Scan will scan the image
+func (s *ClamScanner) Scan(path string, image *docker.Image) error {
+	scanner := clamd.NewClamd(s.Socket)
```
Comment (mfojtik) [inline pkg/clamav/scanner.go]: do we want to check if the socket exists and make some user friendly message?

Comment (mfojtik) [inline pkg/clamav/scanner.go]: probably the scanner interface have Validate() method to check if you have all requirements needed for running scan?

File: pkg/clamav/scanner.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+		fmt.Printf("%v %v\n", s, err)
+	}
+
+	//// TODO result channel
+	//_, err := scanner.ContScanFile(path)
+	return nil
+}
+
+// ScannerName is the scanner's name
+func (s *ClamScanner) ScannerName() string {
```
Comment (mfojtik) [inline pkg/clamav/scanner.go]: just Name() ?

File: pkg/clamav/scanner.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	return "ClamAV"
+}
+
+// ResultFileName returns the name of the results file
+func (s *ClamScanner) ResultsFileName() string {
```
Comment (mfojtik) [inline pkg/clamav/scanner.go]: hrm. don't we want to return io.Reader instead?

File: pkg/api/types.go (hunk 2 above)
Comment (mfojtik) [inline pkg/api/types.go]: @enoodle we discussed that, we should first "standardize" what the resource should look like to cover all scan results...

File: pkg/inspector/image-inspector.go (hunk 1 above)
Comment (pweil-) [inline pkg/inspector/image-inspector.go]: Yeah, I did that just for the WIP to save time in the refactor since I was using a dummy file in the scanner.  The goal would be that everything just cares about the file system once it gets to the scanner and how that file system is produced is out of scope.

File: pkg/api/types.go (hunk 2 above)
Comment (pweil-) [inline pkg/api/types.go]: +1 to scan options.  Maybe that should be a pre-req to this PR

File: pkg/inspector/image-inspector.go (hunk 4 above)
Comment (pweil-) [inline pkg/inspector/image-inspector.go]: yeah, another one on my TODO that I removed in this refactor just to lessen the changes.  We can make a separate PR to introduce that (and the switch logic) to reduce the junk in this.

File: pkg/api/types.go (hunk 5 above)
Comment (pweil-) [inline pkg/api/types.go]: I was thinking `ScanType string`

File: pkg/clamav/scanner.go (hunk 6 above)
Comment (pweil-) [inline pkg/clamav/scanner.go]: I believe this actually creates the socket so I'm not sure.  I noticed that when I was running the daemon with the same socket and started my test scan the daemon actually quit because something else started using the socket.  Definitely an implementation detail that needs worked out.

File: pkg/clamav/scanner.go (hunk 7 above)
Comment (pweil-) [inline pkg/clamav/scanner.go]: yeah, that is a hold over from the straight port to an interface.

**Comment by pweil-** [comment]
superseded by https://github.com/openshift/image-inspector/pull/55
